dockerfiles/theia/src/main.js

```diff
@@ -11,46 +11,159 @@
  */
 const fs = require("fs");
 const cp = require("child_process");
+
 const theiaRoot = '/home/theia';
-const theiaPath = theiaRoot + '/package.json';
+const defaultTheiaRoot = '/home/default/theia';
+const gitPluginsRoot = '/tmp/theia-plugins-from-git-repos';
+
+prepareTheia();
+
+/** Rebuilds if needed and runs Theia */
+function prepareTheia() {
+    process.chdir(theiaRoot);
+
+    const extraPlagins = getExtraTheiaPlugins();
+    let currentPlugins = getTheiaPlugins();
+    let newPlugins = getDefaultTheiaPlugins();
+
+    if (Object.keys(currentPlugins).length === 0) {
+        copyDefaultTheiaBuild();
+        currentPlugins = getTheiaPlugins();
+    }
+
+    for (let pluginName in extraPlagins) {
+        newPlugins[pluginName] = extraPlagins[pluginName];
+    }
+
+    if (!isPluginsEqual(currentPlugins, newPlugins)) {
+        console.log('Plugins set changed. Rebuilding Theia...');
+        rebuildTheiaWithNewPluginsAndRun(newPlugins);
+    } else {
+        handleError(callRun());
+    }
+}
+
+function isPluginsEqual(pls1, pls2) {
+    if (Object.keys(pls1).length !== Object.keys(pls2).length) {
+        return false;
+    }
+
+    for (let dep in pls1) {
+        if (pls1[dep] !== pls2[dep]) {
+            return false;
+        }
+    }
+
+    return true;
+}
 
-const defaultTheiaPath = `/home/default/theia`;
-const defaultConfig = require(`${defaultTheiaPath}/package.json`);
+function copyDefaultTheiaBuild() {
+    cp.execSync(`cp -r ${defaultTheiaRoot}/node_modules ${theiaRoot} && cp ${defaultTheiaRoot}/yarn.lock ${defaultTheiaRoot}/package.json ${theiaRoot}`);
```

Comment:
Maybe split this commad:
`cp.execSync(`cp -r ${defaultTheiaRoot}/node_modules ${theiaRoot}`);`
 `cp.execSync(`cp ${defaultTheiaRoot}/yarn.lock ${defaultTheiaRoot}/package.json ${theiaRoot}`);`
It should be a bit more readable. But up to you.

Reply:
I just don't want to spawn one more child process for this